enhancements/baremetal/minimise-baremetal-footprint.md

```diff
@@ -0,0 +1,171 @@
+
+Over recent releases OpenShift has improved support for small-footprint
+deployments, in particular with the compact-clusters enhancement which adds
+full support for 3-node clusters where the masters are schedulable.
+
+This is a particularly useful deployment option for baremetal PoC environments,
+where often the amount of physical hardware is limited, but there is still the
+problem of where to run the installer/bootstrap-VM in this environment.
+
+The current solution for IPI baremetal is to require a 4th bootstrap host,
+which is a machine physically connected to the 3 master nodes, that runs
+the installer and/or the bootstrap VM.  This effectively means the minimum
+footprint is 4 nodes, unless you can temporarily connect a provisioning host
+to the cluster machines.
+
+A similar constraint exists for UPI baremetal deployments, where although a
+3 master cluster is possible, you need to run a 4th bootstrap node somewhere
+for the duration of the initial installation.
+
+Even in larger deployments, it is not recommended to host the bootstrap or
+controlplane services on a host that will later host user workloads, due
+to the risk of e.g stale loadbalancer config resulting in controlplane
+traffic reaching that node, so potentially you always need an additional
+node (which may need to be dedicated per-cluster for production cases).
+
+## Motivation
+
+This proposal outlines a potential approach to avoid the requirement for a
+4th node, leveraging the recent etcd-operator improvements and work to enable
+a live-iso replacement for the bootstrap VM.
+
+### Goals
+
+* Enable clusters to be deployed on baremetal with exactly 3 nodes
+* Avoid the need for additional nodes to run install/bootstrap components
+* Simplify existing IPI baremetal day-1 user experience
+
+### Non-Goals
+
+* Supporting any controlplane topology other than three masters.
+* Supporting deployment of a single master or scaling from such a deployment.
+* Support for pivoting the bootstrap machine to a worker.
+
+## Proposal
+
+### User Stories
+
+As a user of OpenShift, I should be able to install a fully supportable
+3-node cluster in baremetal environments, without the requirement to
+temporarily connect a 4th node to host installer/bootstrap services.
+
+As a large-scale production user with multi-cluster deployments I want to avoid
+dedicated provisioning nodes per-cluster in addition to the controlplane node
+count, and have the ability to redeploy in-place for disaster recovery reasons.
+
+As an existing user of the IPI baremetal platform, I want to simplify my day-1
+experience by booting a live-ISO for the bootstrap services, instead of a
+host with a bootstrap VM that hosts those services.
+
+### Risks and Mitigations
+
+This proposal builds on work already completed e.g etcd-operator improvements
+but we need to ensure any change in deployment topology is well tested and
+fully supported, to avoid these deployments being an unreliable
+corner-case.
+
```

Comment:
What is the impact of this change on the current integration with ACM? That expects to use Hive to run the installer in a pod in the ACM cluster, but it seems that will need to change to run something that attaches the live ISO to one of the hosts instead? I don't think we need to work out all of the details here, but we should at least point out that if we make this change in a way that isn't backwards compatible then we will break the existing ACM integration.

Reply:
@dhellmann that is a good point - I suspect that means that at least short/medium term the existing bootstrap VM solution would still be required, and we have the same question to answer re any ACM integration with the assisted-install flow?

I've been wondering whether this proposal could be simplified by not considering the "run installer on the live-iso" part, and instead prepare a cluster-specific ISO the user can then boot e.g `openshift-install create boostrap-iso` ?

That would still imply two possible install paths though, the bootstrap VM case or the alternative based on creation of a bootstrap ISO without dependencies on libvirt.